src/AppInstallerCommonCore/Yaml.cpp

```diff
@@ -0,0 +1,478 @@
+    {
+        if (m_type == Type::Sequence)
+        {
+            m_sequence = decltype(m_sequence)::value_type{};
+        }
+        else if (m_type == Type::Mapping)
+        {
+            m_mapping = decltype(m_mapping)::value_type{};
+        }
+    }
+
+    void Node::SetScalar(std::string value)
+    {
+        Require(Type::Scalar);
+        m_scalar = std::move(value);
+    }
+
+    bool Node::operator<(const Node& other) const
+    {
+        Require(Type::Scalar);
+        other.Require(Type::Scalar);
+        return this->m_scalar < other.m_scalar;
+    }
+
+    Node& Node::operator[](std::string_view key)
+    {
+        Require(Type::Mapping);
+        auto itrs = m_mapping->equal_range(key);
+
+        if (itrs.first == itrs.second)
+        {
+            return s_globalInvalidNode;
+        }
+
+        Node& result = itrs.first->second;
+
+        THROW_HR_IF(APPINSTALLER_CLI_ERROR_DUPLICATE_MAPPING_KEY, ++itrs.first != itrs.second);
+
+        return result;
+    }
+
+    const Node& Node::operator[](std::string_view key) const
+    {
+        Require(Type::Mapping);
+        auto itrs = m_mapping->equal_range(key);
+
+        if (itrs.first == itrs.second)
+        {
+            return s_globalInvalidNode;
+        }
+
+        const Node& result = itrs.first->second;
+
+        THROW_HR_IF(APPINSTALLER_CLI_ERROR_DUPLICATE_MAPPING_KEY, ++itrs.first != itrs.second);
+
+        return result;
+    }
+
+    Node& Node::operator[](size_t index)
+    {
+        Require(Type::Sequence);
+        return m_sequence.value()[index];
+    }
+
+    const Node& Node::operator[](size_t index) const
+    {
+        Require(Type::Sequence);
+        return m_sequence.value()[index];
+    }
+
+    size_t Node::size() const
+    {
+        switch (m_type)
+        {
+        case Type::Invalid:
+        case Type::None:
+        case Type::Scalar:
+            return 0;
+        case Type::Sequence:
+            return m_sequence->size();
+        case Type::Mapping:
+            return m_mapping->size();
+        }
+
+        THROW_HR(E_UNEXPECTED);
+    }
+
+    const std::vector<Node>& Node::Sequence() const
+    {
+        Require(Type::Sequence);
+        return m_sequence.value();
+    }
+
+    const std::multimap<Node, Node>& Node::Mapping() const
+    {
+        Require(Type::Mapping);
+        return m_mapping.value();
+    }
+
+    void Node::Require(Type type) const
+    {
+        THROW_HR_IF(APPINSTALLER_CLI_ERROR_YAML_INVALID_OPERATION, m_type != type);
+    }
+
+    void Node::Require(Type type1, Type type2) const
+    {
+        THROW_HR_IF(APPINSTALLER_CLI_ERROR_YAML_INVALID_OPERATION, m_type != type1 && m_type != type2);
+    }
+
+    std::string Node::as_dispatch(std::string*) const
+    {
+        return m_scalar;
+    }
+
+    int64_t Node::as_dispatch(int64_t*) const
+    {
+        return std::stoll(m_scalar);
+    }
+
+    bool Node::as_dispatch(bool*) const
+    {
+        if (m_scalar == "true")

```

Comment:
>m_scalar [](start = 12, length = 8)

I think yaml spec recommends case insensitive

Reply:
When I looked at it I didn't see it called out.  I can do case insensitive though.

---
In reply to: [468996962](https://github.com/Microsoft/pkg/pull/535#discussion_r468996962) [](ancestors = 468996962)